### PR TITLE
feat(swap): switch pay/receive assets

### LINKED ISSUE #20339: Switch Pay/Receive parameters
As a user I want o switch pay and receive tokens so I can re-fetch a new swap proposal 

image

### PR DESCRIPTION
**Pull request by briansztamfater**
fixes #20339 

### Summary

This PR adds functionality to the switch button in Setup Swap screen to flip between pay and receive assets. We are replicating same behavior as Desktop as for now we don't support editing the receive field.
- If there is a valid proposal shown to the user, we switch both token and receive amount
- If there is not a swap proposal shown to the user, we keep the same amount in the pay input and just switch the tokens

#### Platforms

- Android
- iOS

##### Functional

- wallet / transactions

### Steps to test

- Open Status
- Login
- Go to wallet
- Select an account
- Select Swap option
- Select a token on Select asset to pay screen
- Enter a valid amount
- Wait for swap proposal to appear
- Tap on the Switch button that is in the middle of pay and receive inputs
- Verify the tokens are flipped correctly

status: ready

### CODE REVIEW

File: src/status_im/contexts/wallet/swap/setup_swap/view.cljs
Comment (ilmotta) [inline src/status_im/contexts/wallet/swap/setup_swap/view.cljs]: Not from this PR, but sharing in any case, `refetch-interval` is defined as a state, but no part in the code relies on it as a reactive state, so it could be a ref.

File: src/status_im/contexts/wallet/swap/events.cljs
Comment (ilmotta) [inline src/status_im/contexts/wallet/swap/events.cljs:401]: Leaning on the side of personal preference :) but is it worth the extra indentation and the usage of an anonymous function just to not repeat `[:wallet :ui :swap]` once?

```clojure
{:db (-> db
         (update-in [:wallet :ui :swap]
                    assoc
                    :asset-to-pay     asset-to-receive
                    :asset-to-receive asset-to-pay
                    :amount           (or receive-amount amount))
         (update-in [:wallet :ui :swap]
                    dissoc
                    :swap-proposal
                    :error-response
                    :loading-swap-proposal?
                    :last-request-uuid))}
```

File: src/status_im/contexts/wallet/swap/events.cljs
Comment (ilmotta) [inline src/status_im/contexts/wallet/swap/events.cljs]: A couple comments about this part :)

1. This is quite hard to read from the inside out. What do you think of using the thread first macro?

```clojure
(when amount-out
  (-> amount-out
      hex/normalize-hex
      native-module/hex-to-number
      (number/convert-to-whole-number receive-token-decimals)
      (number/to-fixed receive-token-decimals)))
```

2. Notice that I used `number/to-fixed`. I think we should avoid `toFixed` because it throws an exception on nil values. We never want unhandled exceptions to happen on the client because this will break the UX.

3. There are too many calls that a dev needs to remember to carefully call in order. There is another place which is similar https://github.com/status-im/status-mobile/blob/783504f33261e229a79ad6aa65fa8ecc7334ee93/src/status_im/contexts/wallet/swap/setup_swap/view.cljs#L171-L180

Better I think would be to have two separate utility functions. Both handle (ignore) nil. One could be called `number/to-fixed` and internally use `remove-trailing-zeroes` and another could be `number/hex->whole`, which would call the whole chain `normalize-hex` + `hex-to-number` + `convert-to-whole-number`.

**Review by ilmotta — COMMENTED** [review]
I won't approve or disapprove, but I was walking by and left a review :)

File: src/status_im/contexts/wallet/swap/events.cljs
Comment (briansztamfater) [reply src/status_im/contexts/wallet/swap/events.cljs]: Nice suggestions @ilmotta, using thread first macro makes the code much cleaner, I will implement them.

File: src/status_im/contexts/wallet/swap/setup_swap/view.cljs
Comment (mohsen-ghafouri) [inline src/status_im/contexts/wallet/swap/setup_swap/view.cljs:103]: for some of these value that do some calculation, like `pay-token-fiat-value`, we should use `use-memo` to memoized value and avoid recalculation between  re-renders.

File: src/status_im/contexts/wallet/swap/setup_swap/view.cljs
Comment (mohsen-ghafouri) [inline src/status_im/contexts/wallet/swap/setup_swap/view.cljs]: it can be defined like on-close outside of the component
```suggestion
      [swap-order-button {:on-press on-swap-order}]
```

**Review by alwx — APPROVED** [review]

**Comment by VolodLytvynenko** [comment]
hi @briansztamfater could you please resolve existing conflicts? thanx!

**Comment by VolodLytvynenko** [comment]
## ISSUE 1: Incorrect Token Displayed as Approved After Switching Pay/Receive
### Steps:
1. Go to the swap page using a token that has not been approved yet.
2. Tap the "Approve" button (e.g., approving 1 DAI).
3. While the approval process is in progress, tap the "Switch Pay/Receive" button.
### Actual Result:
- The token that was initially set to be received is now shown as "approved" (toast is shown and 'approved' message) instead of DAI, even though it hasn’t been approved on the blockchain.

https://github.com/user-attachments/assets/3974f537-d423-4dd4-ab8b-14e589a6313a

-  The user is able to proceed to the swap confirmation page for the token that hasn't been approved in the blockchain that causes an error in the app:
image


check the token approval on:
[Etherscan](https://etherscan.io/tokenapprovalchecker)
[Optimistic Etherscan](https://optimistic.etherscan.io/tokenapprovalchecker)
[Arbiscan](https://arbiscan.io/tokenapprovalchecker)

### Expected Result:
The approval process should apply to the token that was originally selected for approval (DAI in this case). 

Potential Solutions:
- Freeze the "Switch Pay/Receive" button during the approval process to prevent switching tokens mid-approval.

### Devices:
iPhone 11 Pro Max, iOS 17
Pixel 7a, Android 13

**Comment by briansztamfater** [comment]
@VolodLytvynenko thank you very much for testing, just rebased and issue 1 should be fixed now 👍

**Comment by VolodLytvynenko** [comment]
Unfortunately, this issue isn’t consistent, and there are no clear steps to reproduce it. Hopefully, the logs will help. I can only reproduce it on mobile, not on desktop. Could you check if it’s related to [issue 21187](https://github.com/status-im/status-mobile/issues/21187)? If so, issue can be skipped in this PR and addressed in a existing one 

## ISSUE 2: "Something went wrong" Error displayed randomly after tapping the Switch Button

### Steps:
1. Go to the swap page.
2. Enter a value and select tokens for both "Pay" and "Receive."
3. Tap the "Switch Pay/Receive" button a few times.

### Actual Result:
 "Something went wrong" error is randomly displayed after tapping the switch button, interrupting the swap process.

https://github.com/user-attachments/assets/bd2e6aff-e1c6-4401-b407-b0d5e15fee52

### Expected Result:
The token selections should switch smoothly without any error message.

### logs:
[logs (2).zip](https://github.com/user-attachments/files/16932873/logs.2.zip)

**Comment by briansztamfater** [comment]
@VolodLytvynenko from the logs I see this error: `status-proxy-0.error: context canceled, status-proxy-fallback-1.error: context canceled, optimism-archival.rpc.grove.city.error: Post \"https://test.api.status.im/grove/optimism/mainnet/\": context canceled, optimism-archival.rpc.grove.city.error: context canceled`
Not sure why it happens, seems like a go error, but the error message is not descriptive enough. Were you able to check if this is reproducible on Desktop too?

**Comment by briansztamfater** [comment]
@VolodLytvynenko I added a temporary error toast until I implement error handling so we can better know why it is failing. If you can try again the latest build and check if you can reproduce the issue, please let me know what is the error displayed in the toast.

**Comment by VolodLytvynenko** [comment]
> @VolodLytvynenko I added a temporary error toast until I implement error handling so we can better know why it is failing. If you can try again the latest build and check if you can reproduce the issue, please let me know what is the error displayed in the toast.

Hi @briansztamfater here is the shown error
image

**Comment by VolodLytvynenko** [comment]
> @VolodLytvynenko from the logs I see this error: `status-proxy-0.error: context canceled, status-proxy-fallback-1.error: context canceled, optimism-archival.rpc.grove.city.error: Post \"https://test.api.status.im/grove/optimism/mainnet/\": context canceled, optimism-archival.rpc.grove.city.error: context canceled` Not sure why it happens, seems like a go error, but the error message is not descriptive enough. Were you able to check if this is reproducible on Desktop too?

@briansztamfater I can't reproduce this on desktop. And this issue easily reproducible on the mobile

**Comment by VolodLytvynenko** [comment]
Additionally, an extra error toast is displayed, likely due to a broken swap transaction for ERC-20 tokens.https://github.com/status-im/status-desktop/issues/16304, however those errors are not shown in our latest develop

## ISSUE 3: Errors are shown after a transaction is confirmed

### Steps:
1. Go to swap
2. Approve and confirm the transaction
3. take a look at the toast errors

https://github.com/user-attachments/assets/c1c417d0-41fd-44ef-b54c-758bf22d3d15


### Actual result:
"toTokenID is required for Swap" and ":code -32602, :message "invalid argument 0:" after confirmation ERC-20 transaction
image

additionally "JSON-RPC client is unavailable" error is shown after going to the login page

image

### Expected result:
No errors are shown

**Comment by briansztamfater** [comment]
@VolodLytvynenko issue 2 should be fixed now. I tried pressing the switch button about 20 times and haven't seen the error on the last build.
Also issue 3 should be fixed.

**Comment by VolodLytvynenko** [comment]
> @VolodLytvynenko issue 2 should be fixed now. I tried pressing the switch button about 20 times and haven't seen the error on the last build. Also issue 3 should be fixed.

@briansztamfater thank you for fixes. Issue 2 is fixed.
But [issue 3](https://github.com/status-im/status-mobile/pull/21179#issuecomment-2340386718) is still reproducible. "toTokenID is required for Swap" Error Shown after confirming swap transaction and additionally the swap value Is predefined
### Steps:
1. Go to swap
2. Approve and confirm the transaction (in my case I use WBTC 0.0000001-> DAI)
3. take a look at the toast errors
4. Go to swap page again
5. Select the same token that was swapped previously

### Actual result:
- Toast error "toTokenID is required for Swap" messages appear after confirming the transaction.
- The swap value is predefined and remains unchanged from the previous transaction, leading to an incorrect display

https://github.com/user-attachments/assets/c4d183e0-37e3-4160-83c6-e53d50246264


### Logs:
[Status-debug-logs.zip](https://github.com/user-attachments/files/16962812/Status-debug-logs.zip)

**Comment by VolodLytvynenko** [comment]
A bit more context about [issue 3](https://github.com/status-im/status-mobile/pull/21179#issuecomment-2343381816): After the ERC-20 value is approved, it will be shown in the [tool](https://optimistic.etherscan.io/tokenapprovalchecker) as approved (e.g., approved 1 DAI).
image

However, after confirming the transaction, the token still shown as approved on mobile.
On desktop, this record disappears after the transaction is confirmed.

**Comment by briansztamfater** [comment]
@VolodLytvynenko tested again, I was able to find the root cause, and pushed the fix. Let me know how testing goes 🙏

**Comment by VolodLytvynenko** [comment]
hi @briansztamfater thank you for your fix. One additional issue is found

## PR_ISSUE 4: Non-Approved value Is shown as approved if non-approved amount Is entered or token Is switched

### Steps:
1. Go to the swap page using a non-approved ERC-20 token.
2. Approve a specific value (e.g., 0.01 USDC).
3. Enter another non-approved value (e.g., 0.02 USDC) or tap the switch pay/receive button.

### Actual Result:
The non-approved value is incorrectly shown as approved

https://github.com/user-attachments/assets/6eda0f1b-5365-4850-bc9f-82587feba908

### Expected Result:
The non-approved value should not be marked as approved

Devices:
Pixel 7a, Android 13
iPhone 11 Pro Max, IOS 17

**Comment by briansztamfater** [comment]
@VolodLytvynenko issue 4 should be fixed now as well :) Let me know how testing goes

**Comment by VolodLytvynenko** [comment]
> @VolodLytvynenko issue 4 should be fixed now as well :) Let me know how testing goes

Hey @briansztamfater . Thank you for fix. The issue is fixed in case if 'switch' button is tapped. Still [Issue 4](https://github.com/status-im/status-mobile/pull/21179#issuecomment-2345986708) is not fixed fully, and a non-approved value is shown as approved in case if value is increased

### Steps:
1. Go to the swap page using a non-approved ERC-20 token.
2. Approve a specific value (e.g., 0.01 USDC).
3. Enter another non-approved value (e.g., 0.02 USDC)

### Actual Result:
The non-approved value is incorrectly shown as approved

https://github.com/user-attachments/assets/1167975e-e5d3-427b-915c-a59fe8acadc8

**Comment by briansztamfater** [comment]
> > @VolodLytvynenko issue 4 should be fixed now as well :) Let me know how testing goes
> 
> Hey @briansztamfater . Thank you for fix. The issue is fixed in case if 'switch' button is tapped. Still [Issue 4](https://github.com/status-im/status-mobile/pull/21179#issuecomment-2345986708) is not fixed fully, and a non-approved value is shown as approved in case if value is increased
> 
> ### Steps:
> 1. Go to the swap page using a non-approved ERC-20 token.
> 2. Approve a specific value (e.g., 0.01 USDC).
> 3. Enter another non-approved value (e.g., 0.02 USDC)
> 
> ### Actual Result:
> The non-approved value is incorrectly shown as approved
> 
>  trim.mp4

@VolodLytvynenko my fault, I misread the steps and only fixed one cause of the issue. I've just pushed the other fix, hopefully now issue 4 is finally fixed!

**Comment by briansztamfater** [comment]
@VolodLytvynenko I also rebased the PR, but since status-go router was updated, I had to fix new issues that appeared because of new signals being sent from status-go for updating fees, and that was affecting auto-refresh and other issues after confirming transaction. I already pushed the fixes, but unfortunately I cannot guarantee that new issues didn't appear, I did my best to detect the issues by myself and fix them beforehand, so let me know if something else appears :)

**Comment by VolodLytvynenko** [comment]
> @VolodLytvynenko I also rebased the PR, but since status-go router was updated, I had to fix new issues that appeared because of new signals being sent from status-go for updating fees, and that was affecting auto-refresh and other issues after confirming transaction. I already pushed the fixes, but unfortunately I cannot guarantee that new issues didn't appear, I did my best to detect the issues by myself and fix them beforehand, so let me know if something else appears :)

hi @briansztamfater thank you for your work and all updates. Unfortunately  some additional issue are found

## PR_ISSUE 5: Boundary data is shown as approved after auto-refresh on the Swap Page
### Steps:
1. Go to the swap page.
2. Approve a specific value of an ERC-20 token (e.g., 0.00001 BTC on Arbitrum).
3. Enter another value that exceeds the approved amount, but within a boundary (e.g., enter 0.000011 BTC).
4. Wait for the auto-refresh.

### Actual Result:
After the auto-refresh, the unapproved boundary value is shown as approved.

https://github.com/user-attachments/assets/7d5bf159-edc0-415f-80f7-ba4fc21facf2

### Expected Result:
The boundary value should still be displayed as unapproved after the auto-refresh.

### Devices:
Pixel 7a, Android 13
iPhone 11 Pro Max, iOS 17

**Comment by VolodLytvynenko** [comment]
### ISSUE 6: The "approve" button is shown with delay

### Steps:
1. Go to the swap page with a token that requires approval
2. Enter a amount that does not need apporoval
3. Then, enter a value that requires approval.

### Actual Result:
The "Approve" button appears after a noticeable delay. And while it appearing, the user is able to proceeds

https://github.com/user-attachments/assets/a4485200-70d4-4118-b5c5-233bd0d9dacc


### Expected Result:
The review button should not be shown as enabled if new value is entered. The system should process new value to check if this value is approve or not

### Devices:
Pixel 7a, Android 13
iPhone 11 Pro Max, iOS 17

**Comment by briansztamfater** [comment]
@VolodLytvynenko Thanks for testing again! Issues should be fixed. Unfortunately latest status-go changes affected the QA process in the middle of testing / development of this PR, hopefully following PRs won't suffer this much of issues / regressions. Let me know if you find anything else.

**Comment by VolodLytvynenko** [comment]
hi @briansztamfater thank you for your hard work and patient! No issues from my side. PR is tested and ready to be merged

**Comment by briansztamfater** [comment]
@VolodLytvynenko Awesome! That’s great to hear. Could you please update the labels? I don’t want to intervene in the process by doing it myself 😄

**Comment by VolodLytvynenko** [comment]
> @VolodLytvynenko Awesome! That’s great to hear. Could you please update the labels? I don’t want to intervene in the process by doing it myself 😄

@briansztamfater apologies. Just updated

**Pull request merged**
